pkg/container-utils/containerd/containerd.go

```diff
@@ -15,31 +15,185 @@
 package containerd
 
 import (
+	"context"
+	"fmt"
 	"time"
 
-	criclient "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/cri"
+	"github.com/containerd/containerd"
+	tasks "github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/api/types/task"
+	"google.golang.org/grpc"
+
 	runtimeclient "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/runtime-client"
 )
 
 const (
 	DefaultTimeout = 2 * time.Second
 )
 
+const (
+	ContainerNameNerdctl = "nerdctl/name"
+	ContainerNameK8s     = "io.kubernetes.container.name"
+)
+
 type ContainerdClient struct {
-	criclient.CRIClient
+	client *containerd.Client
 }
 
 func NewContainerdClient(socketPath string) (runtimeclient.ContainerRuntimeClient, error) {
 	if socketPath == "" {
 		socketPath = runtimeclient.ContainerdDefaultSocketPath
 	}
 
-	criClient, err := criclient.NewCRIClient(runtimeclient.ContainerdName, socketPath, DefaultTimeout)
+	client, err := containerd.New(socketPath,
+		containerd.WithTimeout(DefaultTimeout),
+		containerd.WithDefaultNamespace("k8s.io"),
+		containerd.WithDialOpts([]grpc.DialOption{
+			grpc.WithBlock(),
+		}),
+	)
 	if err != nil {
 		return nil, err
 	}
 
 	return &ContainerdClient{
-		CRIClient: criClient,
+		client: client,
+	}, nil
+}
+
+func (c *ContainerdClient) Close() error {
+	if c.client != nil {
+		return c.client.Close()
+	}
+	return nil
+}
+
+func (c *ContainerdClient) GetContainers() ([]*runtimeclient.ContainerData, error) {
+	taskResponse, err := c.client.TaskService().List(context.TODO(), &tasks.ListTasksRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]*runtimeclient.ContainerData, len(taskResponse.Tasks))
+	for i, task := range taskResponse.Tasks {
+		ret[i], err = c.ContainerdTaskToContainerData(task)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ret, nil
+}
+
+func (c *ContainerdClient) GetContainer(containerID string) (*runtimeclient.ContainerData, error) {
+	response, err := c.client.TaskService().Get(context.TODO(), &tasks.GetRequest{
+		ContainerID: containerID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	containerData, err := c.ContainerdTaskToContainerData(response.Process)
+	if err != nil {
+		return nil, err
+	}
+	return containerData, nil
+}
+
+func (c *ContainerdClient) GetContainerDetails(containerID string) (*runtimeclient.ContainerDetailsData, error) {
+	response, err := c.client.TaskService().Get(context.TODO(), &tasks.GetRequest{
+		ContainerID: containerID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	proc := response.Process
+
+	containerData, err := c.ContainerdTaskToContainerData(proc)
+	if err != nil {
+		return nil, err
+	}
+
+	containers, err := c.client.Containers(context.TODO(), fmt.Sprintf("id==%s", containerID))
+	if err != nil {
+		return nil, err
+	}
+	if len(containers) != 1 {
+		return nil, fmt.Errorf("expected 1 container with id %q, got %d", containerID, len(containers))
+	}
```

Comment:
It is redundant. I see three calls here to enrich every container. It is too much from my point of view, we need to make it simpler and faster. We don't want to delay the startup during container enrichment.

Reply:
I reduced it to 2 calls.
`.TaskService()` to get the running tasks and then `.Containers(id)` to get the rest of the information.